r/content/specify_data_types_and_schemas.Rmd

```diff
@@ -0,0 +1,250 @@
+to cast to these data types or using a schema which contains them will result in
+an error.
+
+```{r, float_16_conversion, error=TRUE, eval=FALSE}
+oscars <- tibble::tibble(
+  actor = c("Katharine Hepburn", "Meryl Streep", "Jack Nicholson"),
+  num_awards = c(4, 3, 3)
+)
+
+# Set up schema with "num_awards" as float16 which doesn't have an R equivalent
+oscars_schema <- schema(actor = string(), num_awards = float16())
+
+Table$create(oscars, schema = oscars_schema)
+
+```
+
+```{r}
+## Error: Invalid: Cannot convert to Half Float
+```
+
+```{r, test_float_16_conversion, opts.label = "test"}
+test_that("float_16_conversion works as expected", {
+  expect_error(Table$create(oscars, schema = schema(actor = string(), num_awards = float16())))
+})
+```
+
+## Specify data types when creating an Arrow table from an R object
+
+You want to manually specify Arrow data types when converting an object from a data frame to an Arrow object.
+
+### Solution
+
+```{r, use_schema}
+# Set up a tibble to use in this example
+oscars <- tibble::tibble(
+  actor = c("Katharine Hepburn", "Meryl Streep", "Jack Nicholson"),
+  num_awards = c(4, 3, 3)
+)
+
+# Set up schema with "num_awards" as integer
+oscars_schema <- schema(actor = string(), num_awards = int16())
+
+# create arrow Table containing data and schema
+oscars_data_arrow <- Table$create(oscars, schema = oscars_schema)
+
+oscars_data_arrow
+```
+```{r, test_use_schema, opts.label = "test"}
+test_that("use_schema works as expected", {
+  expect_s3_class(oscars_data_arrow, "Table")
+  expect_equal(
+    oscars_data_arrow$schema,
+    oscars_schema
+  )
+})
+```
+
+## Specify data types when reading in files
+
+You want to manually specify Arrow data types when reading in files.
+
+### Solution
+
+```{r, use_schema_dataset}
+# Set up a tibble to use in this example
+oscars <- tibble::tibble(
+  actor = c("Katharine Hepburn", "Meryl Streep", "Jack Nicholson"),
+  num_awards = c(4, 3, 3)
+)
+
+# write dataset to disk
+write_dataset(oscars, path = "oscars_data")
+
+# Set up schema with "num_awards" as integer
+oscars_schema <- schema(actor = string(), num_awards = int16())
+
+# read the dataset in, using the schema instead of inferring the type automatically
+oscars_dataset_arrow <- open_dataset("oscars_data", schema = oscars_schema)
+
+oscars_dataset_arrow
+```
+```{r, test_use_schema_dataset, opts.label = "test"}
+test_that("use_schema_dataset works as expected", {
+  expect_s3_class(oscars_dataset_arrow, "Dataset")
+  expect_equal(oscars_dataset_arrow$schema,
+    oscars_schema
+  )
+})
+```
+```{r, include=FALSE}
+unlink("oscars_data", recursive = TRUE)
+```
+
+### Discussion
+
+If your dataset contains files with different schemas with some overlapping 
+fields which have the same names and types, they will be combined by
+`open_dataset()` automatically.  The code below demonstrates how this works 
+via the `unify_schemas()` function.
```

Comment:
Unfortunately it's actually not fully equivalent. `unify_schemas` will error if types are not the same for fields with the same name, while open_dataset can do some schema evolution (promoting some types). This might be too much a detail to go into here, though (and it's maybe also rather a missing feature in `unify_schemas`)

Reply:
Thanks for pointing this out.  I have opened [ARROW-14190](https://issues.apache.org/jira/browse/ARROW-14190) to discuss this, and removed this section to the cookbook as I agree it's too much detail here.